internal/controller/engine/source.go

```diff
@@ -0,0 +1,189 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+
+	kcache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+)
+
+var _ source.Source = &StoppableSource{}
+
+// NewStoppableSource returns a new watch source that can be stopped.
+func NewStoppableSource(infs cache.Informers, t client.Object, h handler.EventHandler, ps ...predicate.Predicate) *StoppableSource {
+	return &StoppableSource{infs: infs, Type: t, handler: h, predicates: ps}
+}
+
+// A StoppableSource is a controller-runtime watch source that can be stopped.
+type StoppableSource struct {
+	infs cache.Informers
+
+	Type       client.Object
+	handler    handler.EventHandler
+	predicates []predicate.Predicate
+
+	reg kcache.ResourceEventHandlerRegistration
+}
+
+// Start is internal and should be called only by the Controller to register
+// an EventHandler with the Informer to enqueue reconcile.Requests.
+func (s *StoppableSource) Start(ctx context.Context, q workqueue.RateLimitingInterface) error {
+	i, err := s.infs.GetInformer(ctx, s.Type, cache.BlockUntilSynced(true))
+	if err != nil {
+		return errors.Wrapf(err, "cannot get informer for %T", s.Type)
+	}
+
+	reg, err := i.AddEventHandler(NewEventHandler(ctx, q, s.handler, s.predicates...).HandlerFuncs())
+	if err != nil {
+		return errors.Wrapf(err, "cannot add event handler")
+	}
+	s.reg = reg
+
+	return nil
+}
+
+// Stop removes the EventHandler from the source's Informer. The Informer will
+// stop sending events to the source.
+func (s *StoppableSource) Stop(ctx context.Context) error {
+	if s.reg == nil {
+		return nil
+	}
+
+	i, err := s.infs.GetInformer(ctx, s.Type)
+	if err != nil {
+		return errors.Wrapf(err, "cannot get informer for %T", s.Type)
+	}
+
+	return errors.Wrap(i.RemoveEventHandler(s.reg), "cannot remove event handler")
+}
+
+// NewEventHandler creates a new EventHandler.
+func NewEventHandler(ctx context.Context, q workqueue.RateLimitingInterface, h handler.EventHandler, ps ...predicate.Predicate) *EventHandler {
+	return &EventHandler{
+		ctx:        ctx,
+		handler:    h,
+		queue:      q,
+		predicates: ps,
+	}
+}
+
+// An EventHandler converts a controller-runtime handler and predicates into a
+// client-go ResourceEventHandler. It's a stripped down version of
+// controller-runtime's internal TypedEventHandler implementation.
```

Comment:
https://github.com/kubernetes-sigs/controller-runtime/blob/v0.18.0/pkg/internal/source/event_handler.go

Specifically this is the above code, which is internal to controller-runtime, with the generics support removed. I should update this comment to link to it explicitly.

Reply:
Added a comment.